File: crates/sieve-py/python/sieve/contrib/openai.py
# SPDX-License-Identifier: MIT OR Apache-2.0
"""OpenAI SDK convenience wrapper.

Installed via ``pip install sieve-guard[openai]``. The wrapper monkey-patches
``client.chat.completions.create`` to:

1. Call ``scanner.scan_input(system, user)`` on the messages.
2. Raise ``PromptInjectionBlocked`` if the verdict blocks.
3. Forward the request to the underlying client.
4. Call ``scanner.scan_output(system, response_text, canary_state)``.
5. Attach the verdict to the response as ``.sieve``.

The wrapper performs ZERO network I/O of its own (R1). All network calls
happen inside the OpenAI client the user supplied — sieve only sees the
strings that flow through.
"""

from __future__ import annotations

from typing import Any
from sieve import Scanner, PromptInjectionBlocked, instrument_system_prompt


def _extract_messages(kwargs: dict[str, Any]) -> tuple[str, str]:
    """Return (system_prompt, last_user_message) from a `messages` kwarg.

    Falls back to empty strings if the kwarg shape isn't recognized — the
    scanner is robust to empty inputs and will simply emit no findings.
    """
    messages = kwargs.get("messages", []) or []
    system = ""
    user = ""
    for m in messages:
        role = m.get("role") if isinstance(m, dict) else getattr(m, "role", None)
        content = m.get("content") if isinstance(m, dict) else getattr(m, "content", "")
        if not isinstance(content, str):
            # Multi-part content (image+text) — flatten the text parts.
            try:
                content = "".join(
                    p.get("text", "") if isinstance(p, dict) else getattr(p, "text", "")
                    for p in content
                )
            except (TypeError, AttributeError):
                content = ""
        if role == "system":
            system = content
        elif role == "user":
            user = content
    return system, user


def _response_text(resp: Any) -> str:
    try:
        choices = resp.choices if hasattr(resp, "choices") else resp["choices"]
        first = choices[0]
        msg = first.message if hasattr(first, "message") else first["message"]
        content = msg.content if hasattr(msg, "content") else msg["content"]
        return content or ""
    except (AttributeError, KeyError, IndexError, TypeError):
        return ""


def _with_instrumented_system(kwargs: dict[str, Any], instrumented_system: str) -> dict[str, Any]:
    """Return kwargs with the outbound system message replaced or inserted."""
    patched = dict(kwargs)
    messages = list(patched.get("messages", []) or [])
    replaced = False
    out: list[Any] = []

    for m in messages:
        if isinstance(m, dict):
            item = dict(m)
            if item.get("role") == "system":
                item["content"] = instrumented_system
                replaced = True
            out.append(item)
        else:
            if getattr(m, "role", None) == "system":
                try:
                    setattr(m, "content", instrumented_system)
                    replaced = True
                except (AttributeError, TypeError):
                    pass
            out.append(m)

    if not replaced:
        out.insert(0, {"role": "system", "content": instrumented_system})

    patched["messages"] = out
    return patched


def wrap(client: Any, scanner: Scanner | None = None) -> Any:
    """Wrap an OpenAI client. Returns the same client with
    ``chat.completions.create`` monkey-patched in place.

    Args:
        client: An instance of ``openai.OpenAI`` (or anything implementing
            ``client.chat.completions.create(**kwargs)``).
        scanner: Optional pre-built ``sieve.Scanner``. If omitted, a default
            scanner is constructed.

    Returns:
        The same client (mutated). The patched method:
        - sets ``resp.sieve`` on success to the post-flight ``Verdict``.
        - raises ``PromptInjectionBlocked`` on a Block verdict.
    """
    if scanner is None:
        scanner = Scanner()
    original = client.chat.completions.create

    def patched(**kwargs: Any) -> Any:
        system, user = _extract_messages(kwargs)
        pre = scanner.scan_input(system, user)
        if pre.is_block():
            raise PromptInjectionBlocked(pre)
        instrumented_system, canary_state = instrument_system_prompt(system)
        resp = original(**_with_instrumented_system(kwargs, instrumented_system))
        text = _response_text(resp)
        post = scanner.scan_output(system, text, canary_state)
        try:
            setattr(resp, "sieve", post)
        except (AttributeError, TypeError):
            # Frozen / pydantic models: stash on a mutable attribute the
            # caller can still see.
            try:
                resp.__dict__["sieve"] = post  # type: ignore[attr-defined]
            except (AttributeError, TypeError):
                pass
        if post.is_block():
            raise PromptInjectionBlocked(post)
        return resp

    client.chat.completions.create = patched
    return client


__all__ = ["wrap"]

File: packages/nextjs/src/openai.ts
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// OpenAI SDK wrapper (`@sieve/nextjs/openai`).
//
// Usage:
//   import OpenAI from 'openai';
//   import { wrapOpenAI } from '@sieve/nextjs/openai';
//
//   const client = wrapOpenAI(new OpenAI());
//   const resp = await client.chat.completions.create({ ... });
//   console.log(resp.sieve.decision);

import {
  sieveCheck,
  sieveCheckOutput,
  instrumentSystemPrompt,
  PromptInjectionBlocked,
  type Verdict,
} from "./index.js";

interface MessagePart {
  text?: string;
  [key: string]: unknown;
}

interface Message {
  role?: string;
  content?: string | MessagePart[];
  [key: string]: unknown;
}

interface CreateArgs {
  messages?: Message[];
  [key: string]: unknown;
}

interface ChatCompletion {
  choices?: Array<{
    message?: { content?: string | null };
  }>;
  [key: string]: unknown;
}

function extractMessages(args: CreateArgs): { system: string; user: string } {
  const messages = args.messages ?? [];
  let system = "";
  let user = "";
  for (const m of messages) {
    const content = typeof m.content === "string"
      ? m.content
      : Array.isArray(m.content)
        ? m.content.map((p) => p.text ?? "").join("")
        : "";
    if (m.role === "system") system = content;
    else if (m.role === "user") user = content;
  }
  return { system, user };
}

function responseText(resp: ChatCompletion): string {
  return resp.choices?.[0]?.message?.content ?? "";
}

function withInstrumentedSystem(args: CreateArgs, systemPrompt: string): CreateArgs {
  const messages = args.messages ?? [];
  let replaced = false;
  const patched = messages.map((m) => {
    if (m.role === "system") {
      replaced = true;
      return { ...m, content: systemPrompt };
    }
    return m;
  });
  if (!replaced) {
    patched.unshift({ role: "system", content: systemPrompt });
  }
  return { ...args, messages: patched };
}

/**
 * Wrap an OpenAI client so every `chat.completions.create` call is
 * scanned in and out. Returns the same client (mutated in place).
 *
 * - Pre-flight scan runs before the LLM call. A `"Block"` verdict
 *   throws {@link PromptInjectionBlocked} without making the LLM call.
 * - Post-flight scan runs on the assistant message. A `"Block"` verdict
 *   throws {@link PromptInjectionBlocked} after the LLM has been
 *   called — the response is still attached to the error via
 *   `error.verdict`.
 * - Successful responses carry the post-flight verdict on `resp.sieve`.
 */
export function wrapOpenAI<T extends {
  chat: { completions: { create: (...args: any[]) => Promise<any> } };
}>(client: T): T {
  const original = client.chat.completions.create.bind(client.chat.completions);
  client.chat.completions.create = async (args: CreateArgs): Promise<any> => {
    const { system, user } = extractMessages(args);
    const pre = await sieveCheck(system, user);
    if (pre.decision === "Block") {
      throw new PromptInjectionBlocked(pre);
    }
    const instrumented = await instrumentSystemPrompt(system);
    const resp: ChatCompletion = await original(
      withInstrumentedSystem(args, instrumented.system_prompt),
    );
    const text = responseText(resp);
    const post = await sieveCheckOutput(system, text, instrumented.canary_state);
    (resp as { sieve?: Verdict }).sieve = post;
    if (post.decision === "Block") {
      throw new PromptInjectionBlocked(post);
    }
    return resp;
  };
  return client;
}

export { PromptInjectionBlocked };
